[RFC] Windows clipboard support

equalsraf opened the pull request:
Support clipboard provider win32yank for Windows. And reenable the clipboard tests in appveyor.

The win32yank is downloaded and installed with the neovim binary. The binary is 32 bit so no architecture checks are made.

Trying to build on a system where win32yank.exe is not present fails with a fatal error, which is not very kind on developers, but avoids generating broken releases. Perhaps checking for the build type would be ideal.

- I have yet to check in depth, appveyor seems happy and the zip file does contain the binary, for some reason cmake included vcruntime140.dll which I was not expecting at all.
- is it possible the clipboard tests would work without  a working provider?

equalsraf committed:
Windows: Clipboard provider win32yank

equalsraf committed:
third-party: Add helper to fetch binary dependencies

Add cmake script with recipes to download binary dependencies, currently
the only dependency is win32yank.

equalsraf committed:
Windows: Install runtime dependencies with Neovim binary

Look for runtime dependencies diff.exe and win32yank.exe (and
recursively for DLL dependencies) and install them with nvim.exe. If a
dependency is missing a warning will be issued.

equalsraf committed:
Enable clipboard provider tests in Windows

brcolow commented on third-party/cmake/GetWindowsDeps.cmake:
```diff
@@ -0,0 +1,49 @@
+include(CMakeParseArguments)
+
+# This is similar to the build recipes, but instead downloads a third party
+# binaries and installs it under the the DEPS_PREFIX.
```
Typo: Binaries should not singular, not plural (binary)

justinmk commented on runtime/autoload/provider/clipboard.vim:
```diff
@@ -78,6 +78,11 @@ function! provider#clipboard#Executable() abort
+    let s:copy['+'] = 'win32yank -i --crlf'
+    let s:paste['+'] = 'win32yank -o --lf'
+    let s:copy['*'] = s:copy['+']
+    let s:paste['*'] = s:paste['+']
```
should `return 'win32yank'` here

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -292,18 +292,41 @@ target_link_libraries(nvim ${NVIM_EXEC_LINK_LIBRARIES})
+    find_program(BIN_PATH ${BIN})
+    if(NOT BIN_PATH)
+      message(FATAL_ERROR "Unable to find external dependency ${BIN}")
+      continue()
```
is continue() needed after a fatal error?

justinmk commented:
> is it possible the clipboard tests would work without a working provider?

We have some clipboard tests that use a fake provider: https://github.com/neovim/neovim/blob/3bd169c6d5298cb2d96114d4a43ba736657b0937/test/functional/autocmd/textyankpost_spec.lua#
L10-L11

equalsraf commented:
> By "fatal error" you mean the test fails? I think that's ok for now.

No, the check for win32yank in src/nvim/CMakelists.txt causes a fatal error. So the configure phase fails.

> We have some clipboard tests that use a fake provider

I meant the ones from ' test/functional/clipboard/clipboard_provider_spec.lua ' even with the previous errors, the tests did not fail.

raghur commented:
Appveyor build works for me.. no more Checkhealth warnings.

equalsraf commented:
One test is failing in travis but seems unrelated.

justinmk commented on third-party/cmake/GetWindowsDeps.cmake:
```diff
@@ -0,0 +1,49 @@
+  set(URL ${${URL_VNAME}})
+  set(HASH ${${HASH_VNAME}})
+  if(NOT URL OR NOT HASH )
+	  message(FATAL_ERROR "${URL_VNAME} and ${HASH_VNAME} must be set")
```
This error would be confusing. Shouldn't it say `"${_gettool_TARGET}_URL and ${_gettool_TARGET}_SHA256 must be set"`

justinmk replied:
Oh, I think I see: `${URL_VNAME}` _is_ `${_gettool_TARGET}_URL`.

justinmk commented on third-party/cmake/GetWindowsDeps.cmake:
```diff
@@ -0,0 +1,49 @@
+  endif()
+
+  string(TOUPPER "${_gettool_TARGET}_URL" URL_VNAME)
+  string(TOUPPER "${_gettool_TARGET}_SHA256" HASH_VNAME)
```
`URL_VARNAME` and `HASH_VARNAME` might be a bit less cryptic.

justinmk commented on third-party/cmake/GetWindowsDeps.cmake:
```diff
@@ -0,0 +1,49 @@
+  INSTALL_COMMAND ${CMAKE_COMMAND} -E make_directory ${DEPS_INSTALL_DIR}/bin
+    COMMAND ${CMAKE_COMMAND} -E copy 
+             ${DEPS_BUILD_DIR}/src/win32yank/win32yank.exe
+	     ${DEPS_INSTALL_DIR}/bin)
```
Can this logic be part of `GetWindowsDep()`? Then in `third-party/CMakeLists.txt`, after `include(GetWindowsDeps)`, we could call `GetWindowsDep()` there instead of here. That avoids have "win32yank" specified in three different places. That will make things a bit more maintainable as we add more "binary dependencies".

justinmk replied:
Then `GetWindowsDeps.cmake` could be renamed to `GetBinaryDep.cmake`, since it isn't really specific to Windows (though we only will use it there).

justinmk commented on third-party/CMakeLists.txt:
```diff
@@ -174,3 +174,11 @@ add_custom_target(clean-shared-libraries
+
+if(WIN32)
+  set(WIN32YANK_URL https://github.com/equalsraf/win32yank/releases/download/v0.0.2/win32yank.zip)
+  set(WIN32YANK_SHA256 78869bf68565607cda1b6a3d549e2487d59d6f0f16f9b003e123c0086f90309d)
```
These could be argument to `GetWindowsDep()` (aka `GetBinaryDep()`) instead of global variables. That would make more sense because we don't have individual scripts for each binary dep, unlike the `BuildFoo` scripts.

justinmk commented:
> I meant the ones from ' test/functional/clipboard/clipboard_provider_spec.lua ' even with the previous errors, the tests did not fail.

It turns out we also use the fake provider there: https://github.com/equalsraf/neovim/blob/503905d91dda67fdb8adc09f5a7ca3cfc4aa514d/test/functional/clipboard/clipboard_provider_spec.lua#L94

The only tests that require `xclip` are in `test/functional/eval/system_spec.lua`.

equalsraf commented:
@justinmk I moved the URL/HASH to the top of the cmakelists.txt like the others. Renamed the file as GetBinaryDeps.cmake and moved the WIN32 check inside.

justinmk commented on third-party/cmake/GetBinaryDeps.cmake:
```diff
@@ -0,0 +1,51 @@
+endfunction()
+
+if(WIN32)
+  GetBinaryDep(TARGET win32yank
```
I was thinking this should live in `third-party/CMakeLists.txt`. That way `third-party/CMakeLists.txt` remains the single place where we specify third-party nuts and bolts. `GetBinaryDeps.cmake` just provides `GetBinaryDep`.

justinmk commented on third-party/cmake/GetBinaryDeps.cmake:
```diff
@@ -0,0 +1,51 @@
+    INSTALL_COMMAND ${CMAKE_COMMAND} -E make_directory ${DEPS_INSTALL_DIR}/bin
+      COMMAND ${CMAKE_COMMAND} -E copy 
+               ${DEPS_BUILD_DIR}/src/win32yank/win32yank.exe
+  	     ${DEPS_INSTALL_DIR}/bin)
```
All of these binary downloads are going to `src/foo`, right? So can this logic be generalized as part of `GetBinaryDep()` instead of repeating it?

equalsraf commented:
Ok, reworked the GetBinaryDeps bits into place.

justinmk commented:
Thanks @equalsraf! BTW, do you know if there's anything blocking us from generating a `.dmg` file for macOS?
